src/borg/repository.py

```diff
@@ -1207,12 +1236,22 @@ def delete_segment(self, segment):
         if segment in self.fds:
             del self.fds[segment]
         try:
-            os.unlink(self.segment_filename(segment))
+            filename = self.segment_filename(segment)
+            # Truncate segment files before unlink(). This can help a full file system recover.
+            # In this instance (cf. cleanup()) we need to use r+b (=O_RDWR|O_BINARY) and
+            # issue an explicit truncate() to avoid creating a file
+            # if *segment* did not exist in the first place.
+            with open(filename, 'r+b') as fd:
+                fd.truncate()
+            os.unlink(filename)
```

Comment:
if you will need these 2 kinds of deletion code at other places also, it would be nice to have them in a helper function.

maybe they could also be merged into 1 function by providing a sure_exists=False flag additionally to dispatch to the one or the other.